core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1369,7 +1377,11 @@ class ReplicaManager(val config: KafkaConfig,
               val requestLeaderEpoch = partitionState.leaderEpoch
               val requestTopicId = topicIds.get(topicPartition.topic)
 
-              if (!partition.checkOrSetTopicId(requestTopicId)) {
+              val (topicIdOpt, partitionLogOpt) = partition.topicIdAndLog
+              if (!checkTopicId(requestTopicId, topicIdOpt, partitionLogOpt)) {
```

Comment:
Maybe I am missing something, but why is it necessary to set the topicId here? I was expecting that we would do this in `makeLeader` and `makeFollower`. I'm a tad uncomfortable exposing the `Log` object through `Partition` before we have directly associated it.

Reply:
This is in the case of the log already existing without a topic ID. Before I had this, the partition.metadata file was not written on the first LeaderAndIsr request.

Reply:
This is also in the case of restarting the broker.

Reply:
I suppose one option is to directly associate through Partition.createLogIfNotExists when we get or create the partition a few lines earlier. We call this in makeLeaders/makeFollowers as well, so it's no extra work.

Reply:
Hmmm. I guess that makes the log creation in makeLeaders/makeFollowers superfluous. Another option is to add the partition metadata file in another place when we know we have the log associated. We know the log gets associated in `partition.createLogIfNotExists` so maybe there? We check if we have an ID sent to this request + if we don't yet have a topic ID in the log. and assign there? Unfortunately, this is in the raft code path too, but it shouldn't be used since we always already assign topic ID.

Reply:
Yes, I was thinking we would update the topicId in `createLogIfNotExists`.